### DISCUSSION TITLE
Eagerly load additional relationships from user orm object

### DISCUSSION THREAD

**Discussion by jackstar12** [fastapi-users/fastapi-users#964] [General]
Hey!

So, I've tried to think about implementing fastapi-users in my current project but I can't seem to get around this problem:

My user table has some relationships with other tables. Since I'm using async sqlalchemy I need to eagerly load the ones I need whenever I get a user. However, my problem is that when using fastapi-users I can't find a way to realize something like this. I've thought about creating my own DatabaseAdapter but that won't work either because there's no way to pass down any arguments to it.

Does anyone have an idea how to go about this problem?

**Comment by frankie567** [comment]
Hi @jackstar12!

That should be possible by adapting both the SQLAlchemy model and the Pydantic model. So, let's say you have a `projects` relationship on your `UserTable` model. You can add it just as you would do in a regular SQLAlchemy model:

```py
class UserTable(Base, SQLAlchemyBaseUserTable):
    projects = relationship(
        "Project", back_populates="user", cascade="all, delete", lazy="joined"
    )
```

For technical reasons, FastAPI Users will transform everything to `UserDB`, a Pydantic model representing a user in database. So if you define a Pydantic model for `Project`, you can overload `UserDB` and define a `projects` attribute:

```py
class Project(BaseModel):
    id: int
    name: str


class UserDB(User, models.BaseUserDB):
    projects: List[Project]
```

We this pattern, everything should go smoothly.

**Reply by jackstar12** [reply]
Hi, thanks for repyling.

I was already thinking about doing it this way, I think I should clarify my point a bit more:

I don't want to always load all of them because in some situations I only need certain relationships. Because some of them might end up getting quite big, especially if I decide to eagerly load all of them by default I don't want to unnecessarily load all of them from a ressource / performance standpoint. I created wrappers that allow me to load users / tables in such ways that I can specify:

`def some_path(user: User = Depends(CurrentUser(clients=True))):`

Where CurrentUser is a class which then creates a dependency loading the User with the clients relationship.
In order to realize this I would need a way to pass down arguments throughout the UserManager down to the Database implementation, but I haven't found a way to do so.

Another thing I was wondering about:
The documentation states, that passwords are salted and hashed.
However, The UserTable does not have any salt Column. What am I missing?

I really appreciate your efforts!

**Reply by frankie567** [reply]
I see, implementing a dependency wrapper is a nice idea.

The easiest way would be to make another query to the DB with the objects you want to eager-load once you have the user id loaded by FastAPI Users. Admittedly, this will result in 2 queries instead of 1, but it may be acceptable.

I can't figure out a better solution right now, problem being that `current_user` dependency is already highly dynamic, so it's difficult to wrap it in another dependency layer to add another parameter... But I'll think about it.

**Reply by frankie567** [reply]
Regarding password hashing, the default algorithm used is [bcrypt](https://en.wikipedia.org/wiki/Bcrypt), which stores the salt besides the hash, in the same string; that's why we don't need a separate column.

**Reply by jackstar12** [reply]
Yeah, I've also thought how you could go about wrapping a dependency like that.

Since `current_user` is creating a dependency dynamically anyways  maybe something like that could be possible, by just passing down an extra_args dict throughout the process:

``` python
    def current_user(
        self,
        optional: bool = False,
        active: bool = False,
        verified: bool = False,
        superuser: bool = False,
        extra_args: Optional[Dict] = None,
        get_enabled_backends: Optional[EnabledBackendsDependency] = None,
    ):
        signature = self._get_dependency_signature(get_enabled_backends)

        @with_signature(signature)
        async def current_user_dependency(*args, **kwargs):
            user, _ = await self._authenticate(
                *args,
                optional=optional,
                active=active,
                verified=verified,
                superuser=superuser,
                extra_args=extra_args,
                **kwargs,
            )
            return user

        return current_user_dependency
```

after passing it through every layer you could then create dependencies like

```python
fastapi_users.current_users(extra_args=dict(my_relationship=True))
```

But it seems a bit messy and might not be the best approach. What do you think?
